### PR TITLE
[bug]: vllm preemption handling

### PR DESCRIPTION
**Pull request by sammshen**
Previously, Hao Chen (sorry I do not know your github at) identifies that we are not properly loading from LMCache for preempted requests even though we generate a new load spec for them inside of `get_num_new_matched_tokens()` when they are resumed. This PR attempts to fix the problem. 

1. Need to test on some workload that will generate preemption. Any recommendations? 

2. Need also to adapt for `vllm <0.9.2`
@maobaolong @chunxiaozheng any ideas based on the old vllm scheduler output format?

### CODE REVIEW

**Comment by chunxiaozheng** [comment]
hi, @sammshen, the following code is also for compatibility with old VLLM versions, do we need to process with this PR here?

```
        if isinstance(cached_reqs, list):
            for i, req in enumerate(cached_reqs):
                request_tracker = self._request_trackers[req.req_id]
                request_tracker.update(req.new_token_ids, req.new_block_ids)

                req_meta = ReqMeta.from_request_tracker(
                    request_tracker,
                    self._block_size,
                    self._lmcache_chunk_size,
                    load_spec=None,
                    skip_save=force_skip_save,
                    discard_partial_chunks=self._discard_partial_chunks,
                )
                if req_meta is not None:
                    meta.add_request(req_meta)
            return meta
```

**Comment by YaoJiayi** [comment]
@sammshen long output usually gives you preemption

**Comment by cighao** [comment]
@sammshen  In `update_preempted()`, we shouldn’t overwrite `self.token_ids` with `new_token_ids`, since these are just incremental tokens. The tracker already holds the original prompt tokens, so we should append `new_token_ids` to `self.token_ids`, just as we do for regular requests.

Additionally, in `get_num_new_matched_tokens()`, `request.prompt_token_ids` is used to check for KV cache hits, which is appropriate for normal requests. However, for preempted requests, where some output tokens have already been sent to the client, the prefill tokens should include both the original prompt and previously generated output tokens when rescheduling. We may need to use `request.all_token_ids` instead of `request.prompt_token_ids` in this case.

File: lmcache/integration/vllm/vllm_v1_adapter.py [hunk 1]
```diff
@@ -1202,13 +1220,22 @@ def build_connector_meta(
                 )
             new_block_ids = cached_reqs.new_block_ids[i]
 
-            request_tracker.update(new_token_ids, new_block_ids)
+            load_spec = self.load_specs.pop(req_id, None)
```
Comment (maobaolong) [inline lmcache/integration/vllm/vllm_v1_adapter.py]: We need to handle preempted requests for while the `if isinstance(cached_reqs, list)` is True.

**Review by maobaolong — CHANGES_REQUESTED** [review]
@sammshen Left a comment, the < 0.9.2 branch should be handled also

**Comment by sammshen** [comment]
@cighao thanks for your feedback! I completely reimplemented. PTAL

I reviewed vllm side carefully and indeed preempted requests use their previous output tokens like a full prefill.

some other things I made sure to double check and will be grateful for your valuable feedback: 
- preempted requests will try very hard to "get back in" and have to make sure that there are no memory leaks from a massive amount of lookups on lmcache side. this is correct because we unpin on lookup regardless of retrieve.
- update the num_saved_tokens to lmcache_cached_tokens a second time
- for preemption, num tokens to compute may be larger than new token ids
- for preemption, number of input tokens may (most likely will) be larger than the prompt
- for preemption, we need to lookup all tokens and not just the prefill if we save decode cache. this is a preemption unique case. can you guys help take a look? 
- @maobaolong @chunxiaozheng I also refactored the <0.9.2 case tested it with 0.9.1 but got this issue: 
```text
ERROR 08-21 08:12:25 [core.py:517]     self._lmcache_engine.start_load_kv(forward_context, **kwargs)
ERROR 08-21 08:12:25 [core.py:517]   File "/home/ubuntu/samuel/cleanDev/LMCache/lmcache/integration/vllm/vllm_v1_adapter.py", line 695, in start_load_kv
ERROR 08-21 08:12:25 [core.py:517]     assert len(tokens) == len(slot_mapping)
ERROR 08-21 08:12:25 [core.py:517]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 08-21 08:12:25 [core.py:517] AssertionError
```

# Benchmark: 


## Deploy: 

```bash
LMCACHE_SAVE_DECODE_CACHE=True LMCACHE_MAX_LOCAL_CPU_SIZE=100 LMCACHE_USE_EXPERIMENTAL=True vllm serve     meta-llama/Llama-3.1-8B-Instruct     --max-model-len 32000    
 --kv-transfer-config     '{"kv_connector":"LMCacheConnectorV1", "kv_role":"kv_both"}'   --load-format dummy   --disable-log-requests > engine_output.log 2>&1
```

## Preemptive Workload: 
```python
vllm bench serve   --model meta-llama/Llama-3.1-8B-Instruct   --seed 0 --port 8000   --dataset-name random --random-input-len 8000 --random-output-len 4000   --num-prompts 100   --request-rate 16
```

## Correctness Logs
```text
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,753] LMCache INFO:[0m Reqid: cmpl-e5951453df7d471e8aac25957dbc9096-0, Total tokens 8891, LMCache hit tokens: 8704, need to load: 8704 [3m(vllm_v1_adapter.py:1050:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m INFO 08-21 07:17:09 [scheduler.py:495] Preempted request cmpl-e5951453df7d471e8aac25957dbc9096-0 has been resumed!
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,757] LMCache INFO:[0m Reqid: cmpl-d6ea6f87cb0d4f27a01cf8874e4e083b-0, Total tokens 8704, LMCache hit tokens: 8448, need to load: 8448 [3m(vllm_v1_adapter.py:1050:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m INFO 08-21 07:17:09 [scheduler.py:495] Preempted request cmpl-d6ea6f87cb0d4f27a01cf8874e4e083b-0 has been resumed!
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,761] LMCache INFO:[0m Reqid: cmpl-6be7ab18a0954c10841b3c2810368c9c-0, Total tokens 8525, LMCache hit tokens: 8448, need to load: 8448 [3m(vllm_v1_adapter.py:1050:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m INFO 08-21 07:17:09 [scheduler.py:558] Resumed request cmpl-e5951453df7d471e8aac25957dbc9096-0 has 8704 computed tokens
[1;36m(EngineCore_0 pid=3945189)[0;0m INFO 08-21 07:17:09 [scheduler.py:558] Resumed request cmpl-d6ea6f87cb0d4f27a01cf8874e4e083b-0 has 8448 computed tokens
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,787] LMCache INFO:[0m Retrieved 8704 out of 8704 out of total 8704 tokens [3m(cache_engine.py:489:lmcache.v1.cache_engine)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,810] LMCache INFO:[0m Retrieved 8448 out of 8448 out of total 8448 tokens [3m(cache_engine.py:489:lmcache.v1.cache_engine)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,841] LMCache INFO:[0m Storing KV cache for 256 out of 8704 tokens (skip_leading_tokens=8448) for request cmpl-d6ea6f87cb0d4f27a01cf8874e4e083b-0 [3m(vllm_v1_adapter.py:938:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,842] LMCache INFO:[0m Stored 256 out of total 256 tokens. size: 0.0312 gb, cost 1.0970 ms, throughput: 28.4877 GB/s; offload_time: 1.0866 ms, put_time: 0.0104 ms [3m(cache_engine.py:292:lmcache.v1.cache_engine)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,846] LMCache INFO:[0m Reqid: cmpl-6be7ab18a0954c10841b3c2810368c9c-0, Total tokens 8525, LMCache hit tokens: 8448, need to load: 8448 [3m(vllm_v1_adapter.py:1050:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m INFO 08-21 07:17:09 [scheduler.py:495] Preempted request cmpl-6be7ab18a0954c10841b3c2810368c9c-0 has been resumed!
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,850] LMCache INFO:[0m Reqid: cmpl-3e722df7db6f4570a2d53a29db337dd7-0, Total tokens 8354, LMCache hit tokens: 8192, need to load: 8192 [3m(vllm_v1_adapter.py:1050:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m INFO 08-21 07:17:09 [scheduler.py:558] Resumed request cmpl-6be7ab18a0954c10841b3c2810368c9c-0 has 8448 computed tokens
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,874] LMCache INFO:[0m Retrieved 8448 out of 8448 out of total 8448 tokens [3m(cache_engine.py:489:lmcache.v1.cache_engine)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,904] LMCache INFO:[0m Reqid: cmpl-3e722df7db6f4570a2d53a29db337dd7-0, Total tokens 8354, LMCache hit tokens: 8192, need to load: 8192 [3m(vllm_v1_adapter.py:1050:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m INFO 08-21 07:17:09 [scheduler.py:495] Preempted request cmpl-3e722df7db6f4570a2d53a29db337dd7-0 has been resumed!
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,908] LMCache INFO:[0m Reqid: cmpl-7b66de45063848e2a820db93939d8c25-0, Total tokens 8189, LMCache hit tokens: 7936, need to load: 7936 [3m(vllm_v1_adapter.py:1050:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m INFO 08-21 07:17:09 [scheduler.py:558] Resumed request cmpl-3e722df7db6f4570a2d53a29db337dd7-0 has 8192 computed tokens
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,933] LMCache INFO:[0m Retrieved 8192 out of 8192 out of total 8192 tokens [3m(cache_engine.py:489:lmcache.v1.cache_engine)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,959] LMCache INFO:[0m Storing KV cache for 256 out of 9728 tokens (skip_leading_tokens=9472) for request cmpl-4caffe57d33f497d98c143c86c64b737-0 [3m(vllm_v1_adapter.py:938:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,960] LMCache INFO:[0m Stored 256 out of total 256 tokens. size: 0.0312 gb, cost 1.2591 ms, throughput: 24.8192 GB/s; offload_time: 1.2384 ms, put_time: 0.0207 ms [3m(cache_engine.py:292:lmcache.v1.cache_engine)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,964] LMCache INFO:[0m Reqid: cmpl-7b66de45063848e2a820db93939d8c25-0, Total tokens 8189, LMCache hit tokens: 7936, need to load: 7936 [3m(vllm_v1_adapter.py:1050:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m INFO 08-21 07:17:09 [scheduler.py:495] Preempted request cmpl-7b66de45063848e2a820db93939d8c25-0 has been resumed!
[1;36m(EngineCore_0 pid=3945189)[0;0m [32;20m[2025-08-21 07:17:09,968] LMCache INFO:[0m Reqid: cmpl-58a0008fd09e4b7ab28ca93572876194-0, Total tokens 8031, LMCache hit tokens: 7936, need to load: 7936 [3m(vllm_v1_adapter.py:1050:lmcache.integration.vllm.vllm_v1_adapter)[0m
[1;36m(EngineCore_0 pid=3945189)[0;0m INFO 08-21 07:17:09 [scheduler.py:495] Preempted request cmpl-58a0008fd09e4b7ab28ca93572876194-0 has been resumed!
```

**Review by YaoJiayi — APPROVED** [review]
LGTM!

**Comment by YaoJiayi** [comment]
@maobaolong This should be backward compatible.

**Comment by cighao** [comment]
Hi @sammshen . I ran into an issue with the preempted-request tracker update. In `update_request_tracker`, when a request is preempted and later rescheduled, we currently do: `self.token_ids.extend(new_token_ids)`. However, for a preempted request the new_token_ids at resume time are not based on the previously scheduled state. So the tracker should overwrite its token_ids with the scheduler’s new state. I think we should construct the correct new_token_ids for a preempted request in `build_connector_meta`:
```
is_preempted = cached_reqs.resumed_from_preemption[i]
if is_preempted:
    num_tokens_to_compute = (
        request.num_computed_tokens
        + scheduler_output.num_scheduled_tokens[req_id]
    )
    new_token_ids = request.all_token_ids[:num_tokens_to_compute]
```

Then in `request_tracker.update()`, if `is_preempted` we should replace instead of extend:
```
if is_preempted:
    self.token_ids = list(new_token_ids)  # overwrite with current scheduler view
else:
    self.token_ids.extend(new_token_ids)
```

File: lmcache/integration/vllm/vllm_v1_adapter.py [hunk 2]
```diff
@@ -1285,16 +1322,22 @@ def build_connector_meta(
...
+                is_preempted=cached_reqs.resumed_from_preemption[i],
+                lmcache_cached_tokens=lmcache_cached_tokens,
+            )
 
             req_meta = ReqMeta.from_request_tracker(
                 request_tracker,
                 self._block_size,
                 self._lmcache_chunk_size,
-                load_spec=None,
+                load_spec=load_spec,
```
Comment (ziruiliu) [inline lmcache/integration/vllm/vllm_v1_adapter.py:1336]: Change of this line really helps.
I made up a test case that cause preemption, and the preempted request's cache will be corrupted when rescheduled because vllm prefix cache was freed while lmcache does not load here.
I will post an issue about details soon

**Comment by yoo-kumaneko** [comment]
Should we continue with this PR?

**Comment by ApostaC** [comment]
> Should we continue with this PR?
Same ask @sammshen

File: lmcache/integration/vllm/vllm_v1_adapter.py [hunk 3]
```diff
@@ -1056,6 +1067,16 @@ def get_num_new_matched_tokens(
         ):
             return 0
 
+        if request.status == RequestStatus.PREEMPTED and self._save_decode_cache:
+            # if lmcache is saving decode cache,
+            # use all of the token ids accumulated from before preemption
+            # because vllm treats preempted requests as new requests
+            # i.e. essentially a prefill
+            token_ids = torch.tensor(
```
Comment (maobaolong) [inline lmcache/integration/vllm/vllm_v1_adapter.py:1075]: This token_ids will be overrode by the following L1082.

File: lmcache/integration/vllm/vllm_v1_adapter.py [hunk 4]
```diff
@@ -1056,6 +1067,16 @@ def get_num_new_matched_tokens(
         ):
             return 0
 
+        if request.status == RequestStatus.PREEMPTED and self._save_decode_cache:
+            # if lmcache is saving decode cache,
+            # use all of the token ids accumulated from before preemption
+            # because vllm treats preempted requests as new requests
```
Comment (maobaolong) [inline lmcache/integration/vllm/vllm_v1_adapter.py:1073]: Vllm v1 reuse the preempted request?

File: lmcache/integration/vllm/vllm_v1_adapter.py (hunk 3 above)
Comment (sammshen) [reply lmcache/integration/vllm/vllm_v1_adapter.py:1075]: good catch

**Comment by sammshen** [comment]
@maobaolong I am working on this, hopefully will be ready by tomorrow at latest.

**Comment by maobaolong** [comment]
@sammshen As https://github.com/LMCache/LMCache/pull/2007 merged, is this PR still needed?

**Comment by sammshen** [comment]
closed! @cighao feel free to take a look at the new solution :)
